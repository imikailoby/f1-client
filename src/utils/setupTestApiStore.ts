/* eslint-disable @typescript-eslint/no-explicit-any */
import { AnyAction, configureStore, EnhancedStore, Middleware, Reducer } from '@reduxjs/toolkit';

export function setupTestApiStore<
  A extends {
    reducer: Reducer<any, any>;
    reducerPath: string;
    middleware: Middleware;
    util: { resetApiState(): any };
  },
>(api: A) {
  const getStore = () =>
    configureStore({
      reducer: {
        [api.reducerPath]: api.reducer,
      },
      middleware: (gdm) => gdm({ serializableCheck: false, immutableCheck: false }).concat(api.middleware),
    });

  type StoreType = EnhancedStore<
    {
      api: ReturnType<A['reducer']>;
    },
    AnyAction,
    ReturnType<typeof getStore> extends EnhancedStore<any, any, infer M> ? M : never
  >;

  const initialStore = getStore() as StoreType;
  const refObj = {
    api,
    store: initialStore,
  };

  return refObj;
}
