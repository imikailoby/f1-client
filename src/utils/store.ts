import { configureStore } from '@reduxjs/toolkit';
import { driverStandingsApi } from '../api/driverStandings.api';
import { resultsApi } from '../api/results.api';
import { CURRENT_ENV } from '../constants';

const apiMiddlewares = [driverStandingsApi.middleware, resultsApi.middleware];

export const store = configureStore({
  reducer: {
    [driverStandingsApi.reducerPath]: driverStandingsApi.reducer,
    [resultsApi.reducerPath]: resultsApi.reducer,
  },
  devTools: CURRENT_ENV === 'development',
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(apiMiddlewares),
});

export type RootState = ReturnType<typeof store.getState>;
