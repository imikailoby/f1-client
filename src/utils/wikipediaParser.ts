import { WIKI_API_URL } from '../constants';

export const getWikipediaArticleImage = async (url: string): Promise<string | null> =>
  new Promise((resolve) => {
    // eslint-disable-next-line no-useless-escape
    const regexp = new RegExp('[^/]+$');
    const array = regexp.exec(url);
    if (array === null || !array[0]) {
      return resolve(null);
    }

    const fn = array[0];
    const params = `?action=query&origin=*&format=json&prop=pageimages&titles=${fn}&pithumbsize=100`;
    const wikipediaApiUrl = `${WIKI_API_URL}${params}`;
    fetch(wikipediaApiUrl, { method: 'GET' })
      .then(async (res) => {
        const json = await res.json();
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const pages: any[] = Object.values(json.query.pages);
        resolve(pages.length ? pages[0]?.thumbnail?.source : null);
      })
      .catch(() => resolve(null));
  });
