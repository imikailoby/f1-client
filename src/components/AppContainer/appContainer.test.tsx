/**
 * @jest-environment jsdom
 */
import { render } from '@testing-library/react'
import { Provider } from 'react-redux';
import { store } from '../../utils/store';
import { AppContainer } from './';

const component = (
  <Provider store={store}>
    <AppContainer />
  </Provider>
)

describe('<AppContainer />', () => {
  it('renders without crashing', () => {
    const { container } = render(component)
    expect(container).toMatchSnapshot();
  });

  it('should set beforeunload event listener after mount', () => {
    const events: { [key: string]: EventListenerOrEventListenerObject | undefined } = {};
    jest.spyOn(window, 'addEventListener').mockImplementation((event, handle) => {
      events[event] = handle;
    });

    render(component);
    expect(window.addEventListener).toBeCalledWith('beforeunload', expect.any(Function));
  })
  it('should remove beforeunload event listener before unmount', () => {
    const events: { [key: string]: EventListenerOrEventListenerObject | undefined } = {};
    jest.spyOn(window, 'removeEventListener').mockImplementation((event) => {
      events[event] = undefined;
    });

    const wrapper = render(component);
    wrapper.unmount();
    expect(window.removeEventListener).toBeCalledWith('beforeunload', expect.any(Function));
  })
});
