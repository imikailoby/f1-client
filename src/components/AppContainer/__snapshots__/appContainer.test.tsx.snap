// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppContainer /> renders without crashing 1`] = `
<div>
  <header
    class="header"
  >
    <img
      alt="f1 logo"
      loading="lazy"
      src="svg>
"
    />
  </header>
  <div
    class="loader"
  >
    <div
      class="loader"
    >
      <img
        alt="loader"
        src="svg>
"
      />
    </div>
  </div>
  <footer
    class="footer"
  >
    <p>
      F1 Client Â© 2022
    </p>
  </footer>
</div>
`;
